math/arctangent.go

```diff
@@ -1,354 +1,268 @@
+	// note (http:// fredrikj.net/blog/2014/12/faster-high-ctx-Atangents/)
 	// suggests that r= 8 times is a good value for
-	// precision with 1000s of digits to millions
+	// ctx with 1000s of digits to millions
 	// however it was easy to determine there is a
 	// better sliding window to use instead
 	// which is what we use as it turns out
-	// when the precision is large, a larger r value
+	// when the ctx is large, a larger r value
 	// compared to 8 is every effective
-	xFloat64, _ := x.Float64()
-	xFloat64 = stdMath.Abs(xFloat64)
-	//the formula simple but works a bit better then a fix value (8)
-	r := stdMath.Max(0, stdMath.Ceil(0.31554321636851*stdMath.Pow(float64(precision), 0.654095561044508)))
-	p := 1 / stdMath.Pow(2, r)
-	halfed := int(stdMath.Ceil(stdMath.Log(xFloat64/p) / stdMath.Log(2)))
+	xf, _ := x0.Float64()
+	xf = stdMath.Abs(xf)
+	// the formula simple but works a bit better then a fixed value (8)
+	r := stdMath.Max(0, stdMath.Ceil(0.31554321636851*stdMath.Pow(float64(ctx.Precision), 0.654095561044508)))
+	var p float64
+
+	// maxPrec is the largest precision value we can use bit shifts instead of
+	// math.Pow which is more expensive.
+	const maxPrec = 3286
+	if ctx.Precision <= maxPrec {
+		p = 1 / float64(uint64(1)<<uint64(r))
+	} else {
+		p = 1 / stdMath.Pow(2, r)
+	}
+	halved := int(stdMath.Ceil(stdMath.Log(xf/p) / stdMath.Ln2))
 
-	//if the value is already less than 1/(2^r) then halfed
+	// if the value is already less than 1/(2^r) then halfed
 	// will be negative and we don't need to apply
 	// the double angle formula because it would hurt performance
 	// so we'll set halfed to zero
-	if halfed < 0 {
-		halfed = 0
+	if halved < 0 {
+		halved = 0
 	}
 
-	if halfed > 0 {
-		x = decimal.WithPrecision(calculatingPrecision).Copy(x)
+	sq := decimal.WithContext(ctx)
+	for i := 0; i < halved; i++ {
+		ctx.FMA(sq, x0, x0, one)
+		Sqrt(sq, sq)
+		ctx.Add(sq, sq, one)
+		ctx.Quo(x0, x0, sq)
 	}
 
-	for i := 0; i < halfed; i++ {
-		xsq := decimal.WithPrecision(calculatingPrecision).Mul(x, x)
-		x = x.Quo(x, xsq.Add(Sqrt(xsq, xsq.Add(xsq, one)), one))
-	}
-
-	xSquared := decimal.WithPrecision(calculatingPrecision).Mul(x, x)
-	xSqPlus1 := decimal.WithPrecision(calculatingPrecision).Add(xSquared, one)
-	return x, xSquared, xSqPlus1, segment, halfed
+	var x2 decimal.Big
+	ctx.Mul(&x2, x0, x0)
+	var x2p1 decimal.Big
+	ctx.Add(&x2p1, &x2, one)
+	return x0, &x2, &x2p1, segment, halved
 }
 
-func getAtanA(x *decimal.Big) func(n uint64) *decimal.Big {
-	return func(n uint64) *decimal.Big {
-		return x
-	}
-}
-
-func getAtanP(precision int, xSq *decimal.Big) func(n uint64) *decimal.Big {
+func getAtanP(ctx decimal.Context, x2 *decimal.Big) SplitFunc {
+	var p decimal.Big
 	return func(n uint64) *decimal.Big {
+		// P(n) = 2n for all n > 0
 		if n == 0 {
 			return one
 		}
-
-		c2N := decimal.WithPrecision(precision).SetUint64(n)
-		c2N.Mul(c2N, two)
-		c2NXSq := c2N.Mul(c2N, xSq)
-		return c2NXSq
+		if n < stdMath.MaxUint64/2 {
```

Comment:
store stdMath.MaxUint64/2 as  const (do it once) or at the least tdMath.MaxUint64>>1

Reply:
Ha your right. I've been doing java mostly these days which doesn't anything that smart.
In that case keep it as is. It's clearer.